server/src/main/java/org/uiautomation/ios/wkrdp/WebKitNotificationListener.java

```diff
@@ -27,162 +35,278 @@
-          WebkitPage newOne = Collections.max(m.getPages());
-          pages.addAll(m.getPages());
-
-          driver.setPages(pages);
-
-          if (driver.getPages().size() == 0) {
-            //log.fine("first page. Nothing to do.");
-          } else if (newOne.isITunesAd()) {
-            //log.fine("itunes ad - ignoring it.");
-          } else {
-            WebkitPage focus = newOne;
-
-            if (session != null) {
-              waitForWindowSwitchingAnimation();
-              driver.switchTo(focus);
-            } else {
-              driver.switchTo(focus);
-            }
-          }
-          // a page disappeared, the driver needs to know.
-        } else if (change < 0) {
-          List<WebkitPage> old = new ArrayList<WebkitPage>();
-          old.addAll(driver.getPages());
+  private void processAddedPages(ApplicationSentListingMessage applicationSentListingMessage) {
+    List<WebkitPage> pages = new ArrayList<WebkitPage>();
 
-          for (WebkitPage p : m.getPages()) {
-            old.remove(p);
-          }
+    // Keep all the pages currently known to the driver
+    pages.addAll(driver.getPages());
 
-          // TODO freynaud problem here.How to handle a window disappearing without loosing the state
-          // of the currently selected page. driver.getPages() may need to become mutable again.
-          // in the normal case, this is the page with the focus that is closed, so loosing the state
-          // is ok. But for the extra window automatically closed ( ie the "there is an app for that
-          // page, downlaod it on itune" page header, the header disappears when driver.get navigates
-          // to a new page. The header disapeearing shouldn't impact the current state.
-          for (WebkitPage p : old) {
-            log.fine(
-                "the page " + p + " has been deleted and must be removed from the driver cache");
-            int currentFocus = driver.getCurrentPageID();
-            if (p.getPageId() == currentFocus) {
-              log.fine("the page deleted is the one with the focus.");
-            } else {
-              driver.setPages(m.getPages());
-            }
-          }
-        }
+    // Remove all the pages we already know of from the message
+    for (WebkitPage p : driver.getPages()) {
+      applicationSentListingMessage.getPages().remove(p);
+    }
+    if (applicationSentListingMessage.getPages().size() == 0) {
+      throw new WebDriverException(applicationSentListingMessage.getPages().size() + " new pages.");
+    }
 
-        sync.signalSimSentPages();
+    // TODO there can be more than one 'new' UIWebView, picking the max one for now.
+    WebkitPage newOne = Collections.max(applicationSentListingMessage.getPages());
+    pages.addAll(applicationSentListingMessage.getPages());
+    driver.setPages(pages);
+    if (driver.getPages().size() == 0) {
+      // log.fine("first page. Nothing to do.");
```

Comment:
@krmahadevan I have no idea Krishnan, it was carried from earlier code. I does not serve a purpose, I think I can remove it.